chalice/chalicelib/matrix_handler.py

```diff
@@ -0,0 +1,187 @@
+import os
+import tempfile
+import traceback
+import loompy
+
+from abc import ABC, abstractmethod
+from typing import List
+from chalicelib import get_mtx_paths, get_size, rand_uuid, clean_dir
+from chalicelib.config import MERGED_MTX_BUCKET_NAME, TEMP_DIR, s3_blob_store, logger
+from chalicelib.hca_download_worker import HcaDownloadWorker
+from chalicelib.request_handler import RequestHandler, RequestStatus
+
+
+class MatrixHandler(ABC):
+    """
+    A generic matrix handler for matrices concatenation
+    """
+
+    def __init__(self, suffix) -> None:
+        self._suffix = suffix
+
+    @property
+    def suffix(self):
+        return self._suffix
+
+    def _download_mtx(self, bundle_uuids: List[str], temp_dir: str) -> List[str]:
+        """
+        Filter for the matrix files within bundles, and download them locally.
+        :param bundle_uuids: A list of bundle uuids.
+        :param temp_dir: A temporary directory for storing all downloaded matrix files.
+        :return: A list of downloaded matrix files paths.
+        """
+        jobs_pool = list(bundle_uuids)
+        hca_download_workers = dict()
+        parallelism = 4
+
+        # Create and assign worker to work on a job only when needed
+        while len(jobs_pool) > 0 or len(hca_download_workers) > 0:
+            wip_jobs = list(hca_download_workers.keys())
+
+            for wip_job in wip_jobs:
+                hca_download_worker = hca_download_workers[wip_job]
+
+                # Catch the exception happens in the worker
+                if hca_download_worker.exception is not None:
+                    logger.info(f'Exception caught, the size of tmp directory is: {get_size(TEMP_DIR)} bytes.')
+                    logger.info(f'tmp directory contains: {os.listdir(TEMP_DIR)}.')
+                    raise hca_download_worker.exception
+
+                # Kick off the worker from the list when it finishes its work
+                if hca_download_worker.exitcode is not None:
+                    del hca_download_workers[wip_job]
+
+            # Assign jobs to workers until reaching the maximum parallelism or there is no more job left.
+            while len(jobs_pool) > 0 and len(hca_download_workers) < parallelism:
+                job = jobs_pool.pop()
+
+                dest_name = os.path.join(temp_dir, job)
+                keywords = {
+                    "bundle_uuid": job,
+                    "replica": "aws",
+                    "dest_name": dest_name,
+                    "metadata_files": (),
+                    "data_files": (f'*{self.suffix}',)
+                }
+
+                hca_download_workers[job] = HcaDownloadWorker(**keywords)
+                hca_download_workers[job].start()
```

Comment:
This all seems like a lot of code to parallelize a function call over a list. Could you use like `concurrent.futures.ThreadPoolExecutor.map` or `multiprocessing.pool.ThreadPool.map`?

Reply:
That's what I tried to do at the beginning. However, due to the Lambda execution environment not having `/dev/shm` (shared memory for processes) support, I can’t use multiprocessing.Queue or multiprocessing.Pool.

Reply:
Right, but what about the two methods I suggested? You don't have multiple processors in a lambda, and I'm pretty sure you're I/O bound here anyway. So why not use threads?

Reply:
Sounds reasonable. I will try using threads.